Handshake tests refactor

AndrzejKurek opened the pull request:
This PR contains a refactor of SSL tests, to simplify adding new tests and to reduce code duplication, with the future aim being a single handshake function testing buffer resizing in all cases.
This goal is reached by passing a structure with options to the handshake function. These have a default value, so that specialised tests only have to overwrite a couple of them in order to work.
Furthermore, a polymorphic-like mechanism has been introduced to the test cases, so that one calls another, from specialised to more general ones. Thanks to that, dependencies and default values can be provided all on appropriate layers. Also, the .data files are cleaner thanks to that.

Patater commented on tests/suites/test_suite_ssl.data:
```diff
@@ -201,155 +201,160 @@ move_handshake_to_state:MBEDTLS_SSL_IS_SERVER:MBEDTLS_SSL_SERVER_NEW_SESSION_TIC
-handshake:"":MBEDTLS_SSL_MINOR_VERSION_0:MBEDTLS_PK_RSA:"":0:0
+handshake:"":MBEDTLS_SSL_MINOR_VERSION_0:MBEDTLS_PK_RSA:"":0:0:MBEDTLS_SSL_MAX_FRAG_LEN_NONE:100:100:1:1
 
+# Note - the case below will have to updated, since the test sends no data due to a 1n-1 split against BEAST, that was not expected when preparing the fragment counting code.
```
Should we update this now? If not, when? Do we have an issue filed for this?

AndrzejKurek replied:
I propose creating an issue once this PR gets merged, since creating one now would be pointless.

Patater replied:
We can link to the issue from the code if we make the issue now. It wouldn't be a defect with the current solution, more a TODO. And it's a TODO both before and after this PR merges.

AndrzejKurek replied:
Added here: https://github.com/ARMmbed/mbedtls/issues/3070

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,6 +6,39 @@
+void init_handshake_options( handshake_test_options *opts )
+{
+  opts->cipher = "";
+  opts->version = MBEDTLS_SSL_MINOR_VERSION_3;
```
Is this really a good default value for version?

AndrzejKurek replied:
This is a version used by default in ssl_client2 and ssl_server2, so I decided to reuse it here.

Patater replied:
OK. I still question it as a default for us to use in testing, both here and in ssl_client2 and ssl_server2, since a majority of our users probably don't use that version.

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -1556,6 +1589,241 @@ int exchange_data( mbedtls_ssl_context *ssl_1,
 }
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+void perform_handshake( handshake_test_options* options )
```
Should this return an `int`? How do the asserts work in this function? They don't have an "exit" label to go to, like their calling functions do.

AndrzejKurek replied:
`perform_handshake` has an exit label. If a `TEST_ASSERT` fails, it not only goes to exit, but also calls `test_fail`, which is a helper function setting data about the fail in a static (file-scope) variable `test_info_t test_info;` which in turn is checked after each test case.

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3300,245 +3568,128 @@ exit:
-                                               0, &context_buf_len )
-                                             == MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+    /* The goto below is used to avoid an "unused label" warning.*/
+    goto exit;
```
How do we know the handshake went well? Should we assert on `perform_handshake()` return value (assuming we add a return value to that function)?

AndrzejKurek replied:
A global `test_info` structure is checked after each test case.

Patater commented on tests/suites/test_suite_ssl.function:
```diff
@@ -1556,6 +1589,241 @@ int exchange_data( mbedtls_ssl_context *ssl_1,
+    }
+#endif /* MBEDTLS_SSL_RENEGOTIATION */
+
+exit:
```
Ah, there's the exit label. All makes sense now.

piotr-now commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,6 +6,39 @@
+    int expected_srv_fragments;
+    int renegotiate;
+    int legacy_renegotiation;
+} handshake_test_options;
```
We should start adding some simple descriptions of these parameters in the future

AndrzejKurek commented:
I'm currently investigating CI issues.

AndrzejKurek commented:
The only failure was due to the 1n-1 split happening also in the SSLv3 test. 
I changed the number of bytes exchanged between server and client to 0 both in the first commit, merging app data (in the .data file), and in the last, refactor commit, by adding a ` || version == MBEDTLS_SSL_MINOR_VERSION_0` to the if.
I also updated the raised issue.

piotr-now commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3300,373 +3568,129 @@ exit:
-        TEST_ASSERT( mbedtls_endpoint_init( &server, MBEDTLS_SSL_IS_SERVER,
-                                            pk_alg, NULL, NULL, NULL ) == 0 );
+        options.cli_msg_len = 0;
+        options.srv_msg_len = 0;
```
To keep testing the app data maybe we could also use `expected_cli_fragments *= 2;`, but I'm not sure. Anyway, we have this issue added for the future works
